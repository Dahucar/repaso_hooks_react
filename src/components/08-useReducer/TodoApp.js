import React, { useEffect, useReducer } from 'react'
import { FormAddTodo } from './components/FormAddTodo';
import { TodoList } from './components/TodoList';
import { todoReducer } from './todoReducer';

//const initialState = [{
//    id: new Date().getTime(),
//    desc: 'React: practica reducer.',
//    done: false
//}];

const init = () => {
    //traera del localStorage las todos guardadas o bien un arreglo vacio
    // en caso de no tener contenido.
    return JSON.parse( localStorage.getItem('todos') ) || [];
}

export const TodoApp = () => {
    /*
        -> useReducer(reducer, initialState, init)
        reducer: es la funcion que se deberá crear. (creado en otro archivo)
        initialState: es el estado inicial que le daremos.
        init: es una funcion para inicializar el state.

        -> const [state, dispatch]
        dispatch: permite disparar las acciones para que react redibuje los cambios en pantalla.
        state: el estado actual.
    */
    const [ todos, dispatch ] = useReducer(todoReducer, [], init);
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify( todos ));
    }, [ todos ]);
    
    // Elimina elemento del todo list
    const handleDelete = ( todoId ) => {
        // crear la accion
        const deleteTodoReducer = {
            type: 'delete',
            payload: todoId
        }
        // hacer el dispatch
        dispatch( deleteTodoReducer );
    }

    // Actualizar estado de todo.
    const hanldeToggleTodo = ( todoId ) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }

    // añade un nuevo todo -> esta funcion es enviada al FormAddTodo component
    const handleAddTodo = newTodo => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    } 

    return (
        <div className="containe-fluid card">
            <div className="card-header">
                <h2>TodoApp</h2>
            </div>
            <div className="card-body">
                <div className="card">
                    <div className="row">
                        <div className="col-8">
                            <TodoList 
                                todos={ todos } 
                                hanldeToggleTodo={ hanldeToggleTodo }
                                handleDelete={ handleDelete } 
                            />
                        </div>
                        <div className="col-4">
                            <FormAddTodo 
                                handleAddTodo={ handleAddTodo }
                            />
                        </div>
                    </div>
                </div>
            </div>
        </div>
    )
}
