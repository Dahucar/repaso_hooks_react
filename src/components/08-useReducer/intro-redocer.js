// IMPORTANTEN: en React el "push" a la calle.
console.log(' Probando el Reducer ');

const initialState = [
    {
        id: 1,
        todo: 'Comprar pan.',
        done: false
    },
    {
        id: 2,
        todo: 'Comprar azucar.',
        done: false
    },
    {
        id: 3,
        todo: 'Comprar sal.',
        done: false
    }
];

const todoReducer = ( state = initialState, action ) => {
    //mediante la accion veremos que hacer cuando se envie un nuevo estado
    if ( action?.type === 'agregar' ) {
        return [ ...state, action.payload ]; // mediante eñ operador spred añado los elementos anteriores y el nuevo.
    }

    // siempte se debe devolver un estado
    return state;
}

let todos = todoReducer();

// 1. Añadir un nuevo todo a mi state inicial.
// un nuevo OBJ, que de alguna forma se debera enviar al reducer mediante la accion.
const newTodo = {
    id: 4,
    todo: 'Comprar leche.',
    done: false
};

// 2. Crear una accion 
const agregarTodoAction = {
    type: 'agregar', // indicara que se hara cuando se envie.
    payload: newTodo // lo que se envia con la accion
};

// 3. Enviar el nuevo elemento al reducer.
// IMPORTANTE: nunca se envia el objeto tal cual al reducer. 
todos = todoReducer( todos, agregarTodoAction );


console.log( todos );