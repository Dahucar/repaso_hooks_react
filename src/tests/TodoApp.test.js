import React from 'react';
import { mount, shallow } from "enzyme";
import { TodoApp } from "../components/08-useReducer/TodoApp";
import { demoTodos } from '../data/demoTodo';
import { act } from '@testing-library/react';

describe('Pruebas en <TodoApp />', () => {
    const wrapper = shallow( <TodoApp />);
    test('Debe hacer match con la instantanea.', () => {
        expect( wrapper ).toMatchSnapshot(); 
    });
    test('Debe agregar un todo', () => {
        // para probar el componente, este trae info de los hijos y de su contexto de uso 
        // tiene problemas con versiones de React
        const wrapper = mount( <TodoApp /> ); 
        act(() => {
            wrapper.find('FormAddTodo').prop('handleAddTodo')( demoTodos[0] );
            wrapper.find('FormAddTodo').prop('handleAddTodo')( demoTodos[1] );
        });
    });
});
