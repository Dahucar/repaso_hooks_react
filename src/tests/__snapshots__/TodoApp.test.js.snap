// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> Debe hacer match con la instantanea. 1`] = `
<div
  className="containe-fluid card"
>
  <div
    className="card-header"
  >
    <h2>
      TodoApp
    </h2>
  </div>
  <div
    className="card-body"
  >
    <div
      className="card"
    >
      <div
        className="row"
      >
        <div
          className="col-8"
        >
          <TodoList
            handleDelete={[Function]}
            hanldeToggleTodo={[Function]}
            todos={Array []}
          />
        </div>
        <div
          className="col-4"
        >
          <FormAddTodo
            handleAddTodo={[Function]}
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
